WcaOnRails/app/controllers/registrations_controller.rb

```diff
@@ -439,163 +441,246 @@ def register
+  # Code skeleton according to https://stripe.com/docs/webhooks/quickstart
+  def stripe_webhook
+    payload = request.body.read
+
+    begin
+      event = Stripe::Event.construct_from(
+        JSON.parse(payload, symbolize_names: true),
+      )
+    rescue JSON::ParserError => e
+      # Invalid payload
+      logger.warn "Stripe webhook error while parsing basic request. #{e.message}"
+      return head :bad_request
+    end
+    # Check if webhook signing is configured.
+    if EnvVars.STRIPE_WEBHOOK_SECRET.present?
+      # Retrieve the event by verifying the signature using the raw body and secret.
+      signature = request.env['HTTP_STRIPE_SIGNATURE']
+      begin
+        event = Stripe::Webhook.construct_event(
+          payload, signature, EnvVars.STRIPE_WEBHOOK_SECRET
+        )
+      rescue Stripe::SignatureVerificationError => e
+        logger.warn "Stripe webhook signature verification failed. #{e.message}"
+        return head :bad_request
+      end
+    elsif Rails.env.production? && EnvVars.WCA_LIVE_SITE?
+      logger.error "No Stripe webhook secret defined in Production."
+      return head :bad_request
+    end
+
+    # Create a default audit that marks the event as "unhandled".
+    audit_event = StripeWebhookEvent.create_from_api(event)
+
+    stripe_intent = event.data.object # contains a polymorphic type that depends on the event
+    stored_transaction = StripeTransaction.find_by(stripe_id: stripe_intent.id)
+
+    if StripeWebhookEvent::HANDLED_EVENTS.include?(event.type)
+      if stored_transaction.nil?
+        logger.error "Stripe webhook reported event on entity #{stripe_intent.id} but we have no matching transaction."
+        return head :not_found
+      else
+        audit_event.update!(stripe_transaction: stored_transaction, handled: true)
+      end
+    end
+
+    # Handle the event
+    case event.type
+    when StripeWebhookEvent::PAYMENT_INTENT_SUCCEEDED
+      # stripe_intent contains a Stripe::PaymentIntent as per Stripe documentation
+
+      stored_intent = stored_transaction.stripe_payment_intent
+
+      stored_intent.update_status_and_charges(stripe_intent, audit_event) do |charge_transaction|
+        if stored_intent.holder.is_a? Registration # currently, the only holders that we pay for are Registrations.
+          ruby_money = charge_transaction.money_amount
+
+          stored_intent.holder.record_payment(
+            ruby_money.cents,
+            ruby_money.currency.iso_code,
+            charge_transaction,
+            stored_intent.user.id,
+          )
+        end
+      end
+    when StripeWebhookEvent::PAYMENT_INTENT_CANCELED
+      # stripe_intent contains a Stripe::PaymentIntent as per Stripe documentation
+
+      stored_intent = stored_transaction.stripe_payment_intent
+      stored_intent.update_status_and_charges(stripe_intent, audit_event)
+    else
+      logger.info "Unhandled Stripe event type: #{event.type}"
+    end
+
+    head :ok
+  end
+
   def payment_success
```

Comment:
Should we use a different name - eg, "payment_completion"? The name could imply that the payment itself was successful - when in fact this functionn seems to render all possible payment state (

Reply:
Done :)